OCPBUGS-43552: maxUnavailable value is not honored when disabling OCL

djoshy opened the pull request:
<!--
If this is a bug fix, make sure your description includes "Fixes: #xxxx", or
"Closes: #xxxx"

Please provide the following information:
-->

**- What I did**
Added an additional check in `isNodeDone` to verify that a node rolling back is not mistakenly considered as done updating.

**- How to verify it**
As the bug states, this is hard to reproduce, but the typical reproduction path is:

1. Configure maxUnavailable=2 in the worker pool
2. Create a MOSC resource for the worker pool and wait for the images to be applied
3. Remove the MOSC
4. Check if maxUnavailable value is honored

It takes several tries(5+) for this to cause the bug, as it relies on the daemon being slow _enough_ to react to the controller's node annotation updates, it's really a luck of the draw.

<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

djoshy commented:
/test unit
/test verify
/test e2e-gcp-op-ocl

sergiordlr commented:
Using IPI on AWS. We haven't been able to reproduce the issue with this fix. We tried about 20 times using maxUnavailable=2 and 3 nodes, and another 5-10 times using maxUnavailable=7 with 8 nodes. The maxUnavailable value was always honored.

We can see the log lines reporting the nodes being already updated

```
I0130 15:47:45.376230       1 status.go:407] isNodeDone: Rollback of node ip-10-0-2-75.us-east-2.compute.internal in process, node is considered working
I0130 15:47:45.376284       1 status.go:407] isNodeDone: Rollback of node ip-10-0-34-192.us-east-2.compute.internal in process, node is considered working
I0130 15:47:45.415044       1 status.go:407] isNodeDone: Rollback of node ip-10-0-34-192.us-east-2.compute.internal in process, node is considered working
```

There is always the possibility that we were lucky, but I think that we can consider that this fix solves the problem. We can approve the PR once the extra logs are removed. Thank you very much!

yuqi-zhang commented on pkg/controller/node/status.go:
```diff
@@ -398,6 +398,16 @@ func isNodeDone(node *corev1.Node, layered bool) bool {
+	// from layered to non-layered. In those cases, cconfig==dconfig, but the node
+	// will still need to do an update back to dconfig's OSImageURL. We can detect a
+	// rolling back node by checking if the cimage stills exists but the dimage does not exist.
+	if cok && !dok {
```
I think this will help in the general sense. I guess the only thing to be careful about is that we don't get soft stuck in situations where the current image annotation exists only, but neither the daemon nor the controller can make progress. I guess that generally shouldn't happen though?

djoshy replied:
Not in the general case, no. If the daemon is unable to clear its current image annotation, it probably means it is failing to rollback to the original `MachineConfig`'s image, and it'd be a cluster wide issue. We should see a node degrade -> pool degrade, I'd think.

djoshy commented:
I removed all the debug logs, and attempted to add a unit test for this particular scenario. However, I noticed that the existing OCL unit tests for node_controller aren't working(some results seems incorrect) and need additional fixes. Since this bug is high priority, I'll break that off into a separate issue, and add a unit test for the rollback case when I've fixed the existing OCL unit tests.

sergiordlr commented:
/label qe-approved

yuqi-zhang reviewed (approved):
/lgtm
